### PR TITLE
Make ignored relative

### LINKED ISSUE #2991: Store relative path in the compiler stat for ignored modules
**I'm submitting a feature request**

**Webpack version:**
1.13.2

**Please tell us about your environment:**
Windows 10, Windows 7

**Current behavior:**
After including [es6-promise](https://www.npmjs.com/package/es6-promise) to my project I see absolute path inside comiler state json file. It is looks like

```
...
"node_modules\\es6-promise\\dist\\es6-promise.js": 61,
"node_modules\\webpack\\node_modules\\node-libs-browser\\node_modules\\process\\browser.js": 62,
"node_modules\\webpack\\buildin\\module.js": 63,
"ignored D:\\MyProject\\node_modules\\es6-promise\\dist vertx": 64,
...
```

**Expected/desired behavior:**
Relative path for ignored entry.

```
"ignored node_modules\\es6-promise\\dist vertx": 64,
```

**What is the motivation / use case for changing the behavior?**
It causes some problems with version control.

 #1315
I already implemented [rohitlodha's solution](https://github.com/webpack/webpack/issues/1315#issuecomment-243107287), but the problem still there when chunks collection grows. `manifest` chunk always added last. So dependent chunks like `vendor` get a new hash.

### PR DESCRIPTION
**Pull request by timse**
**What kind of change does this PR introduce?**

bugfix

**Did you add tests for your changes?**

Yes

**If relevant, link to documentation update:**

N/A

**Summary**

Based on #3147 which seams to not get updated, fixes #2991

**Does this PR introduce a breaking change?**

No?

**Other information**

is it ok to add a plugin to the webpack config in the test, to create that case? @sokra 
Using webpack.IgnorePlugin or the lake wouldnt trigger a record entry.

### CODE REVIEW

File: test/configCases/records/issue-2991/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	entry: "./test",
+	recordsPath: path.resolve(__dirname, "../../../js/config/records/issue-2991/records.json"),
+	target: "node",
+	node: {
+		__dirname: false
+	},
+	plugins: [
+		{
+			apply(compiler) {
+				compiler.plugin("normal-module-factory", (nmf) => {
```
Comment (sokra) [inline test/configCases/records/issue-2991/webpack.config.js]: Use the IgnorePlugin here.

Comment (timse) [inline test/configCases/records/issue-2991/webpack.config.js]: i tried to, however as the IgnoredPlugin intercepts on 'before-resolve' it never gets passed this point: https://github.com/webpack/webpack/blob/master/lib/NormalModuleFactory.js#L235

therefore never makes it into the factory plugin and does not create a `ignored` entry.

The question of course is if it should do that, however there is no real benefit in it is there?

maybe more general, is there even a need to record ignored modules at all?

File: test/configCases/records/issue-2991/test.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+} catch(e){}
+
+it("should write relative paths to records", function() {
+	var fs = require("fs");
+	var path = require("path");
+	var content = fs.readFileSync(path.join(__dirname, "records.json"), "utf-8");
+	content.replace(/\\\\/g, "/").should.eql(`{
+  "modules": {
+    "byIdentifier": {
+      "../../../../external \\"fs\\"": 0,
```
Comment (sokra) [inline test/configCases/records/issue-2991/test.js]: Why are there so many `../` in front of this? Looks like there is already something that tries to make it relative, but it looks like that this is broken. see `makeRelative` in Records plugin.

File: test/configCases/records/issue-2991/webpack.config.js (hunk 1 above)
Comment (sokra) [inline test/configCases/records/issue-2991/webpack.config.js]: Ah ok.

I think you can cause this ignore with a `browser: { foo: false }` in package.json in a browser build. The tests run with target node, so you may add the browser field with `resolve.aliasFields: [ "browser" ]`.

File: test/configCases/records/issue-2991/test.js (hunk 2 above)
Comment (timse) [inline test/configCases/records/issue-2991/test.js]: happy to look into that, but I would prefer to fix that in a different PR?

File: test/configCases/records/issue-2991/webpack.config.js (hunk 1 above)
Comment (timse) [inline test/configCases/records/issue-2991/webpack.config.js]: will check that tonight! thanks 👍

File: lib/RecordIdsPlugin.js [hunk 3]
```diff
@@ -4,14 +4,23 @@
...
 "use strict";
 
+const path = require("path");
+
+const looksLikeAbsolutePath = (maybeAbsolutePath) => /^(?:[a-z]:\\|\/)/i.test(maybeAbsolutePath);
+
 function makeRelative(compiler, identifier) {
 	const context = compiler.context;
-	return identifier.split("|").map(str => str.split("!").map(str => path.relative(context, str)).join("!")).join("|");
+	return identifier.split("|")
```
Comment (sokra) [inline lib/RecordIdsPlugin.js]: Maybe you can optimize it with `.split(/([|! ])/)`... (magic split regex trick)

Comment (timse) [inline lib/RecordIdsPlugin.js]: yeah, i do not particular like the makeRelative function, however using split with regexp loses the ability to join by whatever caused the split.
if that it not important I can do it, but that may lead to clashes no?

**Comment by marcodejongh** [comment]
Anything I can do to help? Keen to get my hands on this change as it's hurting my project's long term caching. The CI build will calculate new internal moduleId's for the ignored modules and as a result all the hashes change 😢

File: lib/RecordIdsPlugin.js (hunk 3 above)
Comment (sokra) [inline lib/RecordIdsPlugin.js]:
```
> "afefe|fefrg!fwerrg".split(/([|! ])/)
[ 'afefe', '|', 'fefrg', '!', 'fwerrg' ]
```

split with a regex in brackets will keep the splitters... (magic trick)

Comment (timse) [inline lib/RecordIdsPlugin.js]: javascript is madness.

Comment (timse) [inline lib/RecordIdsPlugin.js]: done

**Comment by sokra** [comment]
Thanks

**Comment by sokra** [comment]
fixed by #4500

**Pull request merged**